README.md
# Instrumental Agent

Instrument anything.

## Setup & Usage

Add the gem to your Gemfile.

```sh
gem 'instrumental_agent'
```

Visit [instrumentalapp.com](instrumentalapp.com) and create an account, then  initialize the agent with your API key, found in the Docs section.

```sh
I = Instrumental::Agent.new('YOUR_API_KEY', :enabled => Rails.env.production?)
```

You'll  probably want something like the above, only enabling the agent in production mode so you don't have development and production data writing to the same value. Or you can setup two projects, so that you can verify stats in one, and release them to production in another.

Now you can begin to use Instrumental to track your application.

```sh
I.gauge('load', 1.23)                # value at a point in time

I.increment('signups')               # increasing value, think "events"

I.time('query_time') do              # time a block of code
  post = Post.find(1)
end
I.time_ms('query_time_in_ms') do     # prefer milliseconds?
  post = Post.find(1)
end
```

**Note**: For your app's safety, the agent is meant to isolate your app from any problems our service might suffer. If it is unable to connect to the service, it will discard data after reaching a low memory threshold.

Want to track an event (like an application deploy, or downtime)? You can capture events that are instantaneous, or events that happen over a period of time.

```sh
I.notice('Jeffy deployed rev ef3d6a') # instantaneous event
I.notice('Testing socket buffer increase', 3.days.ago, 20.minutes) # an event with a duration
```

## Backfilling

Streaming data is better with a little historical context. Instrumental lets you  backfill data, allowing you to see deep into your project's past.

When backfilling, you may send tens of thousands of metrics per second, and the command buffer may start discarding data it isn't able to send fast enough. We provide a synchronous mode that will ensure every stat makes it to Instrumental before continuing on to the next.

**Warning**: You should only enable synchronous mode for backfilling data as any issues with the Instrumental service issues will cause this code to halt until it can reconnect.

```sh
I.synchronous = true # every command sends immediately
User.find_each do |user|
  I.increment('signups', 1, user.created_at)
end
```

## Server Stats

Want some general server stats (load, memory, etc.)? Check out the [instrumental_tools](https://github.com/fastestforward/instrumental_tools) gem.

```sh
gem install instrumental_tools
instrument_server
```

## Agent Control

Need to quickly disable the agent? set :enabled to false on initialization and you don't need to change any application code.


## Capistrano Integration

Add `require "instrumental/capistrano"` to your capistrano configuration and your deploys will be tracked by Instrumental.  Add the API token for the project you want to track to by setting the following Capistrano var:

```ruby
set :instrumental_key, "MY_API_KEY"
```

The following configuration will be added:

```ruby
before "deploy", "instrumental:util:deploy_start"
after  "deploy", "instrumental:util:deploy_end"
before "deploy:migrations", "instrumental:util:deploy_start"
after  "deploy:migrations", "instrumental:util:deploy_end"
after  "instrumental:util:deploy_end", "instrumental:record_deploy_notice"
```

## Tracking metrics in Resque jobs (and Resque-like scenarios)

If you plan on tracking metrics in Resque jobs, you will need to explicitly cleanup after the agent when the jobs are finished.  You can accomplish this by adding `after_perform` and `on_failure` hooks to your Resque jobs.  See the Resque [hooks documentation](https://github.com/defunkt/resque/blob/master/docs/HOOKS.md) for more information.

You're required to do this because Resque calls `exit!` when a worker has finished processing, which bypasses Ruby's `at_exit` hooks.  The Instrumental Agent installs an `at_exit` hook to flush any pending metrics to the servers, but this hook is bypassed by the `exit!` call; any other code you rely that uses `exit!` should call `I.cleanup` to ensure any pending metrics are correctly sent to the server before exiting the process.

## Using with Ruby Enterprise Edition

Users of Ruby Enterprise Edition should plan on using version 0.9.5 of the Instrumental Agent or greater. Please see the [REE wiki page](https://github.com/fastestforward/instrumental_agent/wiki/Using-with-Ruby-Enterprise-Edition) for more information.


## Troubleshooting & Help

We are here to help. Email us at [support@instrumentalapp.com](mailto:support@instrumentalapp.com), or visit the [Instrumental Support](https://fastestforward.campfirenow.com/6b934) Campfire room.


CHANGELOG.md
### 3.0.0 [August 23, 2019]
* Update ffi gem and setup script

### 3.0.0.alpha [August 22, 2019]
* Drop support for outdated versions of Ruby
* Explicitly add support for new versions of Ruby
* Better handling of SSL errors when connecting to Instrumental
* Note: the agent API has NOT changed.  This is a major release because of the significant changes in Ruby versions officially supported.

### 2.1.0 [January 19, 2018]
* Add support for capistrano 3

### 2.0.0 [August 21, 2017]
* Add automatic tracking of common application metrics, official release

### 2.0.0.alpha [August 18, 2017]
* Add automatic tracking of common application metrics

### 1.0.1 [July 12, 2016]
* Make agent initialization threadsafe

### 1.0.0 [May 13, 2016]
* No significant code changes

### 0.13.4 [October 6, 2015]
* Fixes for when agent times out communicating with server

### 0.13.3 [August 18, 2015]
* Fixes for when agent is disconnected from server

### 0.13.2 [August 17, 2015]
* DNS resolution fixes to prevent fork'd child deadlocks

### 0.13.1 [August 3, 2015]
* Revert encrypted support

### 0.13.0 [August 3, 2015]
* Add support for encrypted transport, tests for same

### 0.12.7 [August 11th, 2014]
* Fix MRI 1.8.7 incompatibility due to RUBY_ENGINE constant

### 0.12.6 [August 11th, 2014]
* Send more agent information in hello command
* Don't require celluloid on Ruby versions that it will never support

### 0.12.5 [September 19th, 2013]
* Set license in gemspec

### 0.12.4 [September 19th, 2013]
* Instrumental doesn't officially support IPV6, prefer IPV4

### 0.12.3 [April 26th, 2013]
* Default collector is now collector.instrumentalapp.com

### 0.12.2 [February 28th, 2013]
* Allow customization of Capistrano deploy message

### 0.12.1 [July 30th, 2012]
* Hide unnecessary logging for normal/common exceptions
* LICENSE

### 0.12.0 [July 30th, 2012]
* Add timeout to socket flush, fixes rare issue on REE/Linux
* Send only one buffer full warning
* Agent instances use global logger more consistently
* Minor code cleanups
* Remove rack-middleware

### 0.11.1 [July 19th, 2012]
* Make error messages easily locatable in logs

### 0.11.0 [July 18th, 2012]
* Allow passing count to increment and gauge calls for pre-aggregated values

### 0.10.1 [July 16th, 2012]
* Fix issue with Etc not being required

### 0.10.0 [July 6th, 2012]
* Remove test mode

### 0.9.11 [July 6th, 2012]
* Allow at_exit handler to be called manually for better Resque integration
* Improved error logging

### 0.9.10 [June 29th, 2012]
* Fix flush command when there's nothing to flush
* Support system_timer and SystemTimer gems.

### 0.9.6 [April 5th, 2012]
* Documentation on reliable collection in Resque jobs
* Fix for dead lock issuew

### 0.9.5 [March 23rd, 2012]
* Defer startup of agent thread until metrics are submitted - this update is strongly recommended for anyone using Ruby Enterprise Edition in concert w/ a preforking application server (like Phusion Passenger).
* Add .stop method for cancelling agent processing
* Changes to how defaults are processed at initialization
* Documentation for usage w/ Resque and Resque like scenarios

### 0.9.1 [March 6th, 2012]
* No longer install system_timer on Ruby 1.8.x, but warn if it's not installed

### 0.9.0 [February 20th, 2012]
* Added manual synchronous flushing command
* Fixed bug with data dropping on short-lived forks

### 0.8.3 [February 9th, 2012]
* Removing symbol to proc use for compatibility with older version of Ruby

### 0.8.2 [January 17, 2012]
* Fixing data loss issue when collector was not responding appropriately

### 0.8.1 [January 13, 2012]
* Event timing works when timed events throw exceptions

### 0.8.0 [January 10, 2012]
* Initial support for timing events.

### 0.7.2 [January 5, 2012]
* Deploy durations tracking fixed.

### 0.7.1 [January 5, 2012]
* Support for exponentially encoded float values

### 0.7 [January 2, 2012]
* .notice added to API to allow tracking project-wide events on graphs
* Added Capistrano recipe contributed by [janxious] from (https://github.com/expectedbehavior/)
* Removed Rack middleware
* Logs to STDERR instead of /dev/null
* Synchronous mode can be specified in agent initialization contributed by [janxious] from (https://github.com/expectedbehavior/)
* Added CHANGELOG :)

### 0.6.1 [December 16, 2011]
* Documentation cleanup

### 0.6 [December 13, 2011]
* Synchronous agent support to allow blocking send of metrics
* Message buffer increased to 5000
* Code cleanup

### 0.5.1 [December 12, 2011]
* instrument_server moved to instrumental_tools gem (https://github.com/instrumental/instrumental_tools)

### 0.5 [December 9, 2011]
* Allow negative numbers to be submitted
* agent logger now can be configured per instance
* Better RSpec formatting for tests

### 0.4 [December 1, 2011]
* Support reconnecting on fork() for forking servers like Passenger

### 0.3 [November 17, 2011]
* Support for test_mode on agent to cause submitted metrics to be thrown away when it reaches Instrumental servers
* Fix version not being sent in agent hello
* Exceptions in agent get swallowed and reported to logger

### 0.2 [November 15, 2011]
* Refactored agent to use TCPSocket instead of EventMachine

### 0.1.6 [November 14, 2011]
* Middleware doesn't automatically inject itself into Rack middleware stack

### 0.1.5 [November 1, 2011]
* Documentation changes

### 0.1.4 [November 1, 2011]
* Documentation changes

### 0.1.3 [October 31, 2011]
* Rename the watch_server command to instrument_server
* Documentation changes

### 0.1.2 [October 30, 2011]
* Add enabled flag to control agent use in test/non-development environments
* watch_server runs every 10s

### 0.1.1 [October 27, 2011]
* Linux support for watch_server
* Documentation changes

### 0.1 [October 27, 2011]
* Initial version


spec/spec_helper.rb
$: << File.join(File.dirname(__FILE__), "..", "lib")

require 'instrumental_agent'
require 'test_server'

RSpec.configure do |config|

  config.before(:all) do
  end

  config.after(:all) do
  end

end


def parse_constant(constant)
  constant = constant.to_s
  parts = constant.split("::")
  constant_name = parts.pop
  source = parts.join("::")
  [source.constantize, constant_name]
end

def with_constants(constants, &block)
  saved_constants = {}
  constants.each do |constant, val|
    source_object, const_name = parse_constant(constant)

    saved_constants[constant] = source_object.const_get(const_name)
    Kernel::silence_warnings { source_object.const_set(const_name, val) }
  end

  begin
    block.call
  ensure
    constants.each do |constant, val|
      source_object, const_name = parse_constant(constant)

      Kernel::silence_warnings { source_object.const_set(const_name, saved_constants[constant]) }
    end
  end
end
alias :with_constant :with_constants

class String
  # From Rails
  def constantize
    names = split('::')
    names.shift if names.empty? || names.first.empty?

    constant = Object
    names.each do |name|
      constant = constant.const_defined?(name) ? constant.const_get(name) : constant.const_missing(name)
    end
    constant
  end
end

module Kernel
  # File activesupport/lib/active_support/core_ext/kernel/reporting.rb, line 10
  def silence_warnings
    with_warnings(nil) { yield }
  end

  def with_warnings(flag)
    old_verbose, $VERBOSE = $VERBOSE, flag
    yield
  ensure
    $VERBOSE = old_verbose
  end
end

spec/test_server.rb
class TestServer
  attr_accessor :host, :port, :connect_count, :commands

  def initialize(options={})
    default_options = {
      :listen => true,
      :authenticate => true,
      :response => true,
    }
    @options = default_options.merge(options)

    @connect_count = 0
    @connections = []
    @commands = []
    @host = 'localhost'
    @main_thread = nil
    @client_threads = []
    @fd_to_thread = {}
    listen if @options[:listen]
  end


  def listen
    @port ||= 10001
    @server = TCPServer.new(port)
    @main_thread = Thread.new do
      begin
        # puts "listening"
        loop do
          client = @server.accept
          @connections << client
          @connect_count += 1
          @fd_to_thread[client.to_i] = Thread.new(client) do |socket|
            # puts "connection received"
            loop do
              begin
                command = ""
                while (c = socket.read(1)) != "\n"
                  command << c unless c.nil?
                end
                if !command.empty?
                  # puts "got: #{command}"
                  commands << command
                  if %w[hello authenticate].include?(command.split(' ')[0])
                    if @options[:response]
                      if @options[:authenticate]
                        socket.puts "ok"
                      else
                        socket.puts "gtfo"
                      end
                    end
                  end
                end
              rescue Exception => e
                break
              end
            end
          end
          @client_threads << @fd_to_thread[client.to_i]
        end
      rescue Exception => err
        unless @stopping
          puts "EXCEPTION:", err unless @stopping
          retry
        end
      end
    end
    # puts "server up"
  rescue Exception => err
    # FIXME: doesn't seem to be detecting failures of listen
    puts "failed to get port"
    puts err.message
    @port += 1
    retry
  end

  def host_and_port
    "#{host}:#{port}"
  end

  def stop
    @stopping = true
    disconnect_all
    @main_thread.kill if @main_thread
    @main_thread = nil
    @client_threads.each { |thread| thread.kill }
    @client_threads = []
    begin
      @server.close if @server
    rescue Exception => e
    end
  end

  def disconnect_all
    @connections.each { |c| 
      if (thr = @fd_to_thread[c.to_i])
        thr.kill
      end
      @fd_to_thread.delete(c.to_i)
      c.close rescue false 
    }
    @connections = []
  end
end


lib/instrumental/metric_buffer.rb
module Instrumental
  class MetricBuffer
    attr_accessor :resolution, :delegate

    def initialize(resolution, max_size, delegate)
      @resolution = resolution
      @max_size = max_size
      @delegate = delegate
      @metrics = Hash.new(0)
    end

    def increment(metric, value, time)
      key = ['increment', metric, quantize_time(time)]
      if @metrics.size >= @max_size && !@metrics.include?(key)
        flush_one_metric
      end
      @metrics[key] += value
    end

    def gauge(metric, value, time)
      key = ['gauge', metric, quantize_time(time)]
      if @metrics.size >= @max_size && !@metrics.include?(key)
        flush_one_metric
      end
      @metrics[key] = value
    end

    def quantize_time(time)
      time - time % @resolution
    end

    def flush_one_metric
      keys = @metrics.keys
      key = keys[rand(keys.size)]
      flush_metric(key)
    end

    def flush_metric(key)
      value = @metrics.delete(key)
      type, metric, quantized_time = *key
      @delegate.store(type, metric, value, quantized_time)
    end

    def flush!
      @metrics.keys.each do |key|
        flush_metric(key)
      end
    end
  end
end


lib/instrumental/capistrano/capistrano2.rb
require 'capistrano'
require 'instrumental_agent'
require 'etc'

if Capistrano::Configuration.instance
  Capistrano::Configuration.instance.load do
    namespace :instrumental do
      namespace :util do
        desc "marker for beginning of deploy"
        task :deploy_start do
          set :instrumental_deploy_start, Time.now
        end

        desc "marker for end of deploy"
        task :deploy_end do
          set :instrumental_deploy_end, Time.now
        end
      end

      desc "send a notice to instrumental about the deploy"
      task :record_deploy_notice do
        start_at = exists?(:instrumental_deploy_start) ? instrumental_deploy_start : Time.now
        end_at = exists?(:instrumental_deploy_end) ? instrumental_deploy_end : start_at
        deploy_duration_in_seconds = end_at - start_at
        deployer = Etc.getlogin.chomp
        agent_options = { :synchronous => true }
        agent_options[:collector] = instrumental_host if exists?(:instrumental_host)
        agent = Instrumental::Agent.new(instrumental_key, agent_options)
        message = if exists?(:deploy_message)
          deploy_message
        else
          "#{deployer} deployed #{current_revision}"
        end
        agent.notice(message,
                     start_at,
                     deploy_duration_in_seconds)
        logger.info("Notified Instrumental of deployment")
      end
    end

    before "deploy", "instrumental:util:deploy_start"
    after  "deploy", "instrumental:util:deploy_end"
    before "deploy:migrations", "instrumental:util:deploy_start"
    after  "deploy:migrations", "instrumental:util:deploy_end"
    after  "instrumental:util:deploy_end", "instrumental:record_deploy_notice"
  end
end


lib/instrumental/agent.rb
require 'instrumental/version'
require 'instrumental/system_timer'
require 'logger'
require 'openssl' rescue nil
require 'resolv'
require 'thread'
require 'socket'
require 'metrician'


module Instrumental
  class Agent
    BACKOFF                            = 2.0
    CONNECT_TIMEOUT                    = 20
    EXIT_FLUSH_TIMEOUT                 = 5
    HOSTNAME                           = Socket.gethostbyname(Socket.gethostname).first rescue Socket.gethostname
    MAX_BUFFER                         = 5000
    MAX_RECONNECT_DELAY                = 15
    REPLY_TIMEOUT                      = 10
    RESOLUTION_FAILURES_BEFORE_WAITING = 3
    RESOLUTION_WAIT                    = 30
    RESOLVE_TIMEOUT                    = 1


    attr_accessor :host, :port, :synchronous, :queue, :dns_resolutions, :last_connect_at
    attr_reader :connection, :enabled, :secure

    def self.logger=(l)
      @logger = l
    end

    def self.logger
      if !@logger
        @logger = Logger.new(STDERR)
        @logger.level = Logger::WARN
      end
      @logger
    end

    # Sets up a connection to the collector.
    #
    #  Instrumental::Agent.new(API_KEY)
    #  Instrumental::Agent.new(API_KEY, :collector => 'hostname:port')
    def initialize(api_key, options = {})
      # symbolize options keys
      options.replace(
        options.inject({}) { |m, (k, v)| m[(k.to_sym rescue k) || k] = v; m }
      )

      # defaults
      # host:        collector.instrumentalapp.com
      # port:        8001
      # enabled:     true
      # synchronous: false
      # secure:      true
      # verify:      true
      @api_key         = api_key
      @host, @port     = options[:collector].to_s.split(':')
      @host          ||= 'collector.instrumentalapp.com'
      requested_secure = options[:secure] == true
      desired_secure   = options[:secure].nil? ? allows_secure? : !!options[:secure]
      if !allows_secure? && desired_secure
        logger.warn "Cannot connect to Instrumental via encrypted transport, SSL not available"
        if requested_secure
          options[:enabled] = false
          logger.error "You requested secure protocol to connect to Instrumental, but it is not available on this system (OpenSSL is not defined). Connecting to Instrumental has been disabled."
        end
        desired_secure = false
      end
      @secure          = desired_secure
      @verify_cert     = options[:verify_cert].nil? ? true : !!options[:verify_cert]
      default_port     = @secure ? 8001 : 8000
      @port            = (@port || default_port).to_i
      @enabled         = options.has_key?(:enabled) ? !!options[:enabled] : true
      @synchronous     = !!options[:synchronous]
      @pid             = Process.pid
      @allow_reconnect = true
      @dns_resolutions = 0
      @last_connect_at = 0
      @metrician       = options[:metrician].nil? ? true : !!options[:metrician]
      @start_worker_mutex = Mutex.new
      @queue = Queue.new

      setup_cleanup_at_exit if @enabled

      if @metrician
        Metrician.activate(self)
      end
    end

    # Store a gauge for a metric, optionally at a specific time.
    #
    #  agent.gauge('load', 1.23)
    def gauge(metric, value, time = Time.now, count = 1)
      if valid?(metric, value, time, count) &&
          send_command("gauge", metric, value, time.to_i, count.to_i)
        value
      else
        nil
      end
    rescue Exception => e
      report_exception(e)
      nil
    end

    # Store the duration of a block in a metric. multiplier can be used
    # to scale the duration to desired unit or change the duration in
    # some meaningful way.
    #
    #  agent.time('response_time') do
    #    # potentially slow stuff
    #  end
    #
    #  agent.time('response_time_in_ms', 1000) do
    #    # potentially slow stuff
    #  end
    #
    #  ids = [1, 2, 3]
    #  agent.time('find_time_per_post', 1 / ids.size.to_f) do
    #    Post.find(ids)
    #  end
    def time(metric, multiplier = 1)
      start = Time.now
      begin
        result = yield
      ensure
        finish = Time.now
        duration = finish - start
        gauge(metric, duration * multiplier, start)
      end
      result
    end

    # Calls time and changes durations into milliseconds.
    def time_ms(metric, &block)
      time(metric, 1000, &block)
    end

    # Increment a metric, optionally more than one or at a specific time.
    #
    #  agent.increment('users')
    def increment(metric, value = 1, time = Time.now, count = 1)
      if valid?(metric, value, time, count) &&
          send_command("increment", metric, value, time.to_i, count.to_i)
        value
      else
        nil
      end
    rescue Exception => e
      report_exception(e)
      nil
    end

    # Send a notice to the server (deploys, downtime, etc.)
    #
    #  agent.notice('A notice')
    def notice(note, time = Time.now, duration = 0)
      if valid_note?(note)
        send_command("notice", time.to_i, duration.to_i, note)
        note
      else
        nil
      end
    rescue Exception => e
      report_exception(e)
      nil
    end

    # Synchronously flush all pending metrics out to the server
    # By default will not try to reconnect to the server if a
    # connection failure happens during the flush, though you
    # may optionally override this behavior by passing true.
    #
    #  agent.flush
    def flush(allow_reconnect = false)
      queue_message('flush', {
        :synchronous => true,
        :allow_reconnect => allow_reconnect
      }) if running?
    end

    def enabled?
      @enabled
    end

    def connected?
      @socket && !@socket.closed?
    end

    def logger=(logger)
      @logger = logger
    end

    def logger
      @logger || self.class.logger
    end

    # Stopping the agent will immediately stop all communication
    # to Instrumental.  If you call this and submit another metric,
    # the agent will start again.
    #
    # Calling stop will cause all metrics waiting to be sent to be
    # discarded.  Don't call it unless you are expecting this behavior.
    #
    # agent.stop
    #
    def stop
      disconnect
      if @thread
        @thread.kill
        @thread = nil
      end
      if @queue
        @queue.clear
      end
    end

    # Called when a process is exiting to give it some extra time to
    # push events to the service. An at_exit handler is automatically
    # registered for this method, but can be called manually in cases
    # where at_exit is bypassed like Resque workers.
    def cleanup
      if running?
        logger.info "Cleaning up agent, queue size: #{@queue.size}, thread running: #{@thread.alive?}"
        @allow_reconnect = false
        if @queue.size > 0
          queue_message('exit')
          begin
            with_timeout(EXIT_FLUSH_TIMEOUT) { @thread.join }
          rescue Timeout::Error
            if @queue.size > 0
              logger.error "Timed out working agent thread on exit, dropping #{@queue.size} metrics"
            else
              logger.error "Timed out Instrumental Agent, exiting"
            end
          end
        end
      end
    end

    private

    def with_timeout(time, &block)
      InstrumentalTimeout.timeout(time) { yield }
    end

    def valid_note?(note)
      note !~ /[\n\r]/
    end

    def valid?(metric, value, time, count)
      valid_metric = metric =~ /^([\d\w\-_]+\.)*[\d\w\-_]+$/i
      valid_value  = value.to_s =~ /^-?\d+(\.\d+)?(e-\d+)?$/

      return true if valid_metric && valid_value

      report_invalid_metric(metric) unless valid_metric
      report_invalid_value(metric, value) unless valid_value
      false
    end

    def report_invalid_metric(metric)
      increment "agent.invalid_metric"
      logger.warn "Invalid metric #{metric}"
    end

    def report_invalid_value(metric, value)
      increment "agent.invalid_value"
      logger.warn "Invalid value #{value.inspect} for #{metric}"
    end

    def report_exception(e)
      logger.error "Exception of type #{e.class} occurred:\n#{e.message}\n#{e.backtrace.join("\n")}"
    end

    def ipv4_address_for_host(host, port, moment_to_connect = Time.now.to_i)
      self.dns_resolutions  = dns_resolutions + 1
      time_since_last_connect = moment_to_connect - last_connect_at
      if dns_resolutions < RESOLUTION_FAILURES_BEFORE_WAITING || time_since_last_connect >= RESOLUTION_WAIT
        self.last_connect_at = moment_to_connect
        with_timeout(RESOLVE_TIMEOUT) do
          address  = Resolv.getaddresses(host).select { |address| address =~ Resolv::IPv4::Regex }.first
          self.dns_resolutions = 0
          address
        end
      end
    rescue Exception => e
      logger.warn "Couldn't resolve address for #{host}:#{port}"
      report_exception(e)
      nil
    end

    def send_command(cmd, *args)
      cmd = "%s %s\n" % [cmd, args.collect { |a| a.to_s }.join(" ")]
      if enabled?

        start_connection_worker
        if @queue && @queue.size < MAX_BUFFER
          @queue_full_warning = false
          logger.debug "Queueing: #{cmd.chomp}"
          queue_message(cmd, { :synchronous => @synchronous })
        else
          if !@queue_full_warning
            @queue_full_warning = true
            logger.warn "Queue full(#{@queue.size}), dropping commands..."
          end
          logger.debug "Dropping command, queue full(#{@queue.size}): #{cmd.chomp}"
          nil
        end
      else
        logger.debug cmd.strip
      end
    end

    def queue_message(message, options = {})
      if @enabled
        options ||= {}
        if options[:allow_reconnect].nil?
          options[:allow_reconnect] = @allow_reconnect
        end
        synchronous = options.delete(:synchronous)
        if synchronous
          options[:sync_resource] ||= ConditionVariable.new
          @sync_mutex.synchronize {
            @queue << [message, options]
            options[:sync_resource].wait(@sync_mutex)
          }
        else
          @queue << [message, options]
        end
      end
      message
    end

    def wait_exceptions
      classes = [Errno::EAGAIN]
      if defined?(IO::EAGAINWaitReadable)
        classes << IO::EAGAINWaitReadable
      end
      if defined?(IO::EWOULDBLOCKWaitReadable)
        classes << IO::EWOULDBLOCKWaitReadable
      end
      if defined?(IO::WaitReadable)
        classes << IO::WaitReadable
      end
      classes
    end


    def test_connection
      begin
        @socket.read_nonblock(1)
        @failures = 0
      rescue *wait_exceptions
        # no-op
      end
    end

    def start_connection_worker
      # NOTE: We need a mutex around both `running?` and thread creation,
      # otherwise we could create two threads.
      # Return early and queue the message if another thread is
      # starting the worker.
      return if !@start_worker_mutex.try_lock
      begin
        return if running?
        return unless enabled?
        disconnect
        address = ipv4_address_for_host(@host, @port)
        if address
          @pid = Process.pid
          @sync_mutex = Mutex.new
          @failures = 0
          @sockaddr_in = Socket.pack_sockaddr_in(@port, address)
          logger.info "Starting thread"
          @thread = Thread.new do
            run_worker_loop
          end
        end
      ensure
        @start_worker_mutex.unlock
      end
    end

    def send_with_reply_timeout(message)
      @socket.puts message
      with_timeout(REPLY_TIMEOUT) do
        response = @socket.gets
        if response.to_s.chomp != "ok"
          raise "Bad Response #{response.inspect} to #{message.inspect}"
        end
      end
    end

    def open_socket(sockaddr_in, secure, verify_cert)
      sock = Socket.new(Socket::PF_INET, Socket::SOCK_STREAM, 0)
      sock.connect(sockaddr_in)
      if secure
        context = OpenSSL::SSL::SSLContext.new
        if verify_cert
          context.set_params(:verify_mode => OpenSSL::SSL::VERIFY_PEER | OpenSSL::SSL::VERIFY_FAIL_IF_NO_PEER_CERT)
        else
          context.set_params(:verify_mode => OpenSSL::SSL::VERIFY_NONE)
        end
        ssl_socket = OpenSSL::SSL::SSLSocket.new(sock, context)
        ssl_socket.sync_close = true
        ssl_socket.connect
        sock = ssl_socket
      end
      sock
    end

    def run_worker_loop
      @failures = 0
      begin
      logger.info "connecting to collector"
      command_and_args = nil
      command_options = nil
        with_timeout(CONNECT_TIMEOUT) do
          @socket = open_socket(@sockaddr_in, @secure, @verify_cert)
        end
        logger.info "connected to collector at #{host}:#{port}"
        hello_options = {
          "version" => "ruby/instrumental_agent/#{VERSION}",
          "hostname" => HOSTNAME,
          "pid" => Process.pid,
          "runtime" => "#{defined?(RUBY_ENGINE) ? RUBY_ENGINE : "ruby"}/#{RUBY_VERSION}p#{RUBY_PATCHLEVEL}",
          "platform" => RUBY_PLATFORM
        }.to_a.flatten.map { |v| v.to_s.gsub(/\s+/, "_") }.join(" ")

        send_with_reply_timeout "hello #{hello_options}"
        send_with_reply_timeout "authenticate #{@api_key}"

        loop do
          command_and_args, command_options = @queue.pop
          if command_and_args
            sync_resource = command_options && command_options[:sync_resource]
            test_connection
            case command_and_args
            when 'exit'
              logger.info "Exiting, #{@queue.size} commands remain"
              return true
            when 'flush'
              release_resource = true
            else
              logger.debug "Sending: #{command_and_args.chomp}"
              @socket.puts command_and_args
            end
            command_and_args = nil
            command_options = nil
            if sync_resource
              @sync_mutex.synchronize do
                sync_resource.signal
              end
            end
          end
        end
      rescue Exception => err
        allow_reconnect = @allow_reconnect
        case err
        when EOFError
        # nop
        when Errno::ECONNREFUSED, Errno::EHOSTUNREACH, Errno::EADDRINUSE, Timeout::Error, OpenSSL::SSL::SSLError
          # If the connection has been refused by Instrumental
          # or we cannot reach the server
          # or the connection state of this socket is in a race
          # or SSL is not functioning properly for some reason
          logger.error "unable to connect to Instrumental, hanging up with #{@queue.size} messages remaining"
          logger.debug "Exception: #{err.inspect}\n#{err.backtrace.join("\n")}"
          allow_reconnect = false
        else
          report_exception(err)
        end
        if allow_reconnect == false ||
           (command_options && command_options[:allow_reconnect] == false)
          logger.info "Not trying to reconnect"
          @failures = 0
          return
        end
        if command_and_args
          logger.debug "requeueing: #{command_and_args}"
          @queue << command_and_args
        end
        disconnect
        @failures += 1
        delay = [(@failures - 1) ** BACKOFF, MAX_RECONNECT_DELAY].min
        logger.error "disconnected, #{@failures} failures in a row, reconnect in #{delay}..."
        sleep delay
        retry
      ensure
        disconnect
      end
    end

    def setup_cleanup_at_exit
      at_exit do
        cleanup
      end
    end

    def running?
      !@thread.nil? && @pid == Process.pid && @thread.alive?
    end

    def flush_socket(socket)
      socket.flush
    rescue Exception => e
      logger.error "Error flushing socket, #{e.message}"
    end

    def disconnect
      if connected?
        logger.info "Disconnecting..."
        begin
          with_timeout(EXIT_FLUSH_TIMEOUT) do
            flush_socket(@socket)
          end
        rescue Timeout::Error
          logger.info "Timed out flushing socket..."
        end
        @socket.close
      end
    rescue Exception => e
      logger.error "Error closing socket, #{e.message}"
    ensure
      @socket = nil
    end

    def allows_secure?
      defined?(OpenSSL)
    end
  end
end
